client/extensions/woocommerce/components/d3/base/index.js

```diff
@@ -0,0 +1,78 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { select as d3Select } from 'd3-selection';
+
+export class D3Base extends Component {
+	static propTypes = {
+		className: PropTypes.string,
+		drawChart: PropTypes.func.isRequired,
+		getParams: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.updateParams = this.updateParams.bind( this );
+	}
+
+	state = {};
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.updateParams );
+		this.updateParams();
+	}
+
+	componentDidUpdate() {
+		this.draw();
+	}
+
+	componentWillReceiveProps() {
+		this.updateParams();
```

Comment:
Any concerns about performance here with doing a new `draw()` each time props change?

Reply:
This is an interesting question because, yes we'd like to redraw if things change. However, this component accepts the functions used to update, not the updates themselves (such as changed data). I think there is some simplification that can happen and I'll need to make sure a change in the parent component's props are accurately reflected

Reply:
I removed `componentWillReceiveProps` because the base component won't need to react to changing props, which are simply functions to be used later in `componentDidUpdate`.

I confirmed that a re-render upstream causes a re-draw due to `componentDidUpdate`. In other words, if Sparkline gets different data, it will be reflected in the ensuing re-draw.

>Any concerns about performance here

We won't be facing unnecessary re-drawing unless upstream props change, which is the desired effect. Perhaps the consuming component can use `shouldComponentUpdate` to compare data values.

Reply:
>base component won't need to react to changing props

This is actually wrong. `componentWillReceiveProps` needs to trigger a recalculation of the parameters used by the draw function on new data because they are dependent. For example, `d3ScaleLinear` calculations of scales based on data

I've added back `componentWillReceiveProps`

Reply:
I guess it is something we can keep an eye on and do some perf monitoring once the base component is in use in other places to see if it is re-drawing when not expected